tools/virtiostat.py

```diff
@@ -0,0 +1,221 @@
+parser.add_argument("-T", "--timestamp", action="store_true",
+    help="show timestamp on output")
+parser.add_argument("-D", "--debug", action="store_true",
+    help="print BPF program before starting (for debugging purposes)")
+parser.add_argument("--ebpf", action="store_true",
+    help=argparse.SUPPRESS)
+args = parser.parse_args()
+if not args.interval:
+    args.interval = args.duration
+
+# define BPF program
+bpf_text = """
+#ifndef KBUILD_MODNAME
+#define KBUILD_MODNAME "bcc"
+#endif
+#include <linux/virtio.h>
+#include <bcc/proto.h>
+
+#define VIRTIO_MAX_SGS  128
+#define SG_MAX          18
+
+typedef struct virtio_stat {
+    char driver[16];
+    char dev[12];
+    char vqname[12];
+    u32 in_sgs;
+    u32 out_sgs;
+    u64 in_bw;
+    u64 out_bw;
+} virtio_stat_t;
+
+BPF_HASH(stats, u64, virtio_stat_t);
+
+static struct scatterlist *__sg_next(struct scatterlist *sgp)
+{
+    struct scatterlist sg;
+
+    bpf_probe_read_kernel(&sg, sizeof(sg), sgp);
+    if (sg_is_last(&sg))
+        return NULL;
+
+    sgp++;
+
+    bpf_probe_read_kernel(&sg, sizeof(sg), sgp);
+    if (unlikely(sg_is_chain(&sg)))
+        sgp = sg_chain_ptr(&sg);
+
+    return sgp;
+}
+
+static u64 count_len(struct scatterlist **sgs, unsigned int num)
+{
+    u64 length = 0;
+    unsigned int i, n;
+    struct scatterlist *sgp = NULL;
+
+    for (i = 0; (i < VIRTIO_MAX_SGS) && (i < num); i++) {
+        for (n = 0, sgp = sgs[i]; sgp && (n < SG_MAX); sgp = __sg_next(sgp)) {
+            length += sgp->length;
+            n++;
+        }
+    }
+
+    return length;
+}
+
+static void record(struct virtqueue *vq,
+                   struct scatterlist **sgs,
+                   unsigned int out_sgs,
+                   unsigned int in_sgs)
+{
+    virtio_stat_t newvs = {0};
+    virtio_stat_t *vs;
+    u64 key = (u64)vq;
+
+    vs = stats.lookup(&key);
+    if (!vs) {
+        bpf_probe_read_kernel(newvs.driver, sizeof(newvs.driver), vq->vdev->dev.driver->name);
+        bpf_probe_read_kernel(newvs.dev, sizeof(newvs.dev), vq->vdev->dev.kobj.name);
+        bpf_probe_read_kernel(newvs.vqname, sizeof(newvs.vqname), vq->name);
+        newvs.in_sgs = in_sgs;
+        newvs.out_sgs = out_sgs;
+        newvs.in_bw = count_len(sgs, in_sgs);
+        newvs.out_bw = count_len(sgs, out_sgs);
+        stats.update(&key, &newvs);
+    } else {
+        vs->in_sgs += in_sgs;
+        vs->out_sgs += out_sgs;
+        vs->in_bw += count_len(sgs, in_sgs);
+        vs->out_bw += count_len(sgs, out_sgs);
+    }
+}
+
+int trace_virtqueue_add_sgs(struct pt_regs *ctx,
+                            struct virtqueue *_vq,
+                            struct scatterlist **sgs,
+                            unsigned int out_sgs,
+                            unsigned int in_sgs,
+                            void *data,
+                            gfp_t gfp)
+
+{
+    unsigned int _out_sgs = PT_REGS_PARM3(ctx);
+    unsigned int _in_sgs = PT_REGS_PARM4(ctx);
```

Comment:
why you need the above two lines to get _out_sgs and _in_sgs? They are in argument list and you should be able to use out_sgs and in_sgs directly?